modopt/opt/algorithms/admm.py

```diff
@@ -0,0 +1,343 @@
+        self._mu_new = self.xp.copy(mu)
+        self._mu_old = self.xp.copy(mu)
+
+    def _update(self):
+        self._u_new = self._opti_H(
+            init=self._u_old,
+            obs=self.B.op(self._v_old) + self._u_old - self.b,
+        )
+        tmp = self.A.op(self._u_new)
+        self._v_new = self.solver2(
+            init=self._v_old,
+            obs=tmp + self._u_old - self.c,
+        )
+
+        self._mu_new = self._mu_old + (tmp + self.B.op(self._v_new) - self.b)
+
+        # update cycle
+        self._u_old = self.xp.copy(self._u_new)
+        self._v_old = self.xp.copy(self._v_new)
+        self._mu_old = self.xp.copy(self._mu_new)
+
+        # Test cost function for convergence.
+        if self._cost_func:
+            self.converge = self.any_convergence_flag()
+            self.converge |= self._cost_func.get_cost(self._u_new, self.v_new)
+
+    def iterate(self, max_iter=150):
+        """Iterate.
+
+        This method calls update until either convergence criteria is met or
+        the maximum number of iterations is reached.
+
+        Parameters
+        ----------
+        max_iter : int, optional
+            Maximum number of iterations (default is ``150``)
+        """
+        self._run_alg(max_iter)
+
+        # retrieve metrics results
+        self.retrieve_outputs()
+        # rename outputs as attributes
+        self.u_final = self._u_new
+        self.v_final = self._v_new
+
+    def get_notify_observers_kwargs(self):
+        """Notify observers.
+
+        Return the mapping between the metrics call and the iterated
+        variables.
+
+        Returns
+        -------
+        dict
+           The mapping between the iterated variables
+        """
+        return {
+            'u_new': self._u_new,
+            'v_new': self._v_new,
+            'idx': self.idx,
+        }
+
+    def retrieve_outputs(self):
+        """Retrieve outputs.
+
+        Declare the outputs of the algorithms as attributes: x_final,
+        y_final, metrics.
+        """
+        metrics = {}
+        for obs in self._observers['cv_metrics']:
+            metrics[obs.name] = obs.retrieve_metrics()
+        self.metrics = metrics
+
+
+class FastADMM(ADMM):
+    r"""Fast ADMM Optimisation Algorihm.
+
+    This class implement the fast ADMM algorithm
+    (Algorithm 8 from :cite:`Goldstein2014`)
+
+    Parameters
+    ----------
+    A : OperatorBase
+        Linear operator for u
+    B : OperatorBase
+        Linear operator for v
+    b : array_like
+        Constraint vector
+    solver1 : function
+        Solver for the x update, takes init_value and obs_value as argument.
+        ie, return an estimate for:
+        .. math:: u_{k+1} = \argmin H(u) + \frac{\tau}{2}\|A u - y\|^2
+    solver2 : function
+        Solver for the z update, takes init_value and obs_value as argument.
+        ie return an estimate for:
+        .. math:: v_{k+1} = \argmin G(v) + \frac{\tau}{2}\|Bv - y \|^2
```

Comment:
I think we have used `callable` fairly consistently throughout the code, so we should stick to it.